LoRaWAN: Adding acquisition of metadata, backoff and a cancel_send() API

hasnainvirk opened the pull request:
### Description

This PR adds new APIs to our Mbed LoRaWAN stack which facilitate the application
in acquiring metadata relating to transmission or reception as well as an API to cancel
an outstanding outgoing message.  

Dependency: https://github.com/ARMmbed/mbed-os/pull/6901

Target: Mbed OS 5.9

Internal ref: IOTCELL-575

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

Commit:
Adding acquisition of TX Meta-data

An API is added to fetch any meta-data available after a succesful
transmission. The stack will make the meta data available after the
TX interrupt is processed. User can get the tx meta data after receiving
TX_DONE event.

Commit:
Adding acquisition of RX meta data

User can now inquire about any RX meta data available after a
successful reception.

Commit:
Adding acquisition of backoff time value

This API enables the application to get hold of remaining time after which
the transmission will take place. User can query the backoff time whenever
there is a packet in the TX pipe. If the event for the backoff expiry is
already queued, the stack does not provide backoff metadata.

Commit:
Adding ability to cancel outgoing transmission

Application can use cancel_sending() API to stop any outstanding, outgoing
transmission (a TX which is not already queued for transmission). This can
potentially enable use cases where the application could cancel a transmission
and go to sleep if the backoff period is long enough rather than waiting for
the transmission to happen.

Commit:
Adding useful trace and setting no. of retries

MCPS confirmation should be filled with the current number of retries
if the ack is not recieved for a CONFIRMED message.

Ack retry number needs to be incremented after the retry is made not before
that.

A few traces are added at the crucial junctions of the code just to tally with
the conceptual flow for debug purposes.

hasnainvirk commented:
@kivaisan @kjbracey-arm @0xc0170 @cmonr @janjongboom Please review.

0xc0170 commented:
@hasnainvirk can you look at astyle issues - see travis astyle? You can ignore some files that are already on master (will be fixed separately). but for new code, would be nice to align

kivaisan commented on features/lorawan/LoRaWANStack.cpp:
```diff
@@ -252,6 +257,22 @@ lorawan_status_t LoRaWANStack::enable_adaptive_datarate(bool adr_enabled)
+{
+     if (_loramac.clear_tx_pipe() == LORAWAN_STATUS_OK) {
+         if (_device_current_state == DEVICE_STATE_SENDING &&
+             _loramac.get_device_class() == CLASS_A) {
```
Why only in class A?

Commit:
Doing away with rx_abort()

rx_abort() was sort of dead code as it was doing essentially nothing.
It might have actually meddled with the state machine if it was hit by
invoking on_ack_timeout_timer_event().

State machine and corresponding processors now take care of the ack timeout,
retries and all other bits, so we don't need abort_rx().

kivaisan commented on features/lorawan/lorawan_types.h:
```diff
@@ -366,4 +367,60 @@ typedef struct lora_channelplan {
+    /**
+     * The transmission time on air of the frame.
+     */
+    uint32_t tx_toa;
```
Please reorder variables to avoid padding

kivaisan commented on features/lorawan/lorawan_types.h:
```diff
@@ -366,4 +367,60 @@ typedef struct lora_channelplan {
+     * The SNR for the received packet.
+     */
+    uint8_t snr;
+} lorawan_rx_metadata;
```
Please reorder variables to avoid padding

kivaisan commented on features/lorawan/lorawan_types.h:
```diff
@@ -366,4 +367,60 @@ typedef struct lora_channelplan {
+    /**
+     * Frame pending status.
+     */
+    uint8_t fpending_status;
```
Probably fpending bit is not needed. If auto uplink is disabled, stack sends UPLINK_REQUIRED event.

Commit:
Adding ack expiry handling for class C

In Class C, rx timeout does not take place for RX2 windows, so if we have
not received anything, we would be retrying but if the no. of retries are
maxed out, and we have not recieved anything yet, we need a mechanism to
tell the upper layer that this has happened.

Commit:
Travis astyle corrections

The scope of style corrections is local to this PR only.

hasnainvirk commented:
@0xc0170 Style corrections for remaining item could be the topic of another PR. I rectified only those style conflicts pertaining to this PR.

hasnainvirk commented:
@kjbracey-arm Can you please review again ? I added a few commits.

kjbracey commented:
If you're basing on #6901, going to have to wait for its review before any merge of this is possible.

hasnainvirk commented:
@kjbracey-arm reviewing doesn't hurt :) I know we have to wait for #6901

Commit:
Adding checks for stack initialization

Metadata APIs should return an error if the stack is not initialized yet.

hasnainvirk commented:
@0xc0170 This needs CI. Failed for some unrelated issue in the previous run.

cmonr commented:
@hasnainvirk We can start CI once this is rebased. Some PRs recently came into master, causing conflicts.

A rebase will also relaunch pr-head, whose issues have been resolved earlier today.

cmonr commented:
@kjbracey-arm @kivaisan Mind doing a quick re-review whilst CI is being dealt with?

cmonr commented:
/morph build

cmonr commented:
@deepikabhavnani Can you help out with this failure? NRF51 in IAR is failing the stats_cpu test.

It's possible NRF51 changes were merged in around the same time that the CPU test was, which had led to an odd conflict.

cmonr commented:
Stopping the export job. It was going to be requeued due to some job shuffling, but since this will need a fix, might as well stop it.

hasnainvirk commented:
@cmonr I held this PR on purpose. I am testing some instability. I will update this PR no later than today.

Commit:
Stability improvements & CRYPTO_ERROR addition

General stability improvements are performed.
A flag is added if a Class C RX2 window is open.
We shouldn't open it again if its already opened.

TX_CRYPTO_ERROR is renamed to CRYPTO_ERROR.
Keeping TX_CRYPTO_ERROR for backwards compatibility.

hasnainvirk commented:
@kivaisan @kjbracey-arm Please review the latest commit. @cmonr Now the CI can be run on this.

cmonr commented:
Going to run since CI is just about empty at the moment. Normally we'd wait until at least one reviewer gives the ok, but both mentioned reviewers were already out but the time this PR was ready for review.

/morph build

kivaisan reviewed (changes requested):
I don't think we need a flag for RX2 window in class C.
As discussed, there are three cases where RX2 is opened:
- When switching from class A to C
- After TX_DONE
- After RX1

So for me it looks like there is a bug if we try to reopen it while it is already open. Please investigate this more.

cmonr commented:
@hasnainvirk Please take a look at @kivaisan's comment.

adbridge commented:
After a discussion with @ChiefBureaucraticOfficer and the fact that Kevin has approved changes and it is unlikely there will be anyone in Finland able to comment on this now before RC1 codefreeze. We will take it in as is.
@hasnainvirk Please consider the comments made by @kivaisan and submit a PR for 5.9-rc2 to fix if required.

hasnainvirk commented:
@adbridge Understood.
